File: frontend/src/lib/copilot-compare.ts
import { z } from "zod";

export const changeDraftSchema = z.object({
  ticker: z.string(),
  amount: z.string(),
  proceeds: z.enum(["cash", "repay_margin"]),
});
export type ChangeDraft = z.infer<typeof changeDraftSchema>;
export const emptyChange: ChangeDraft = { ticker: "", amount: "", proceeds: "cash" };
const receiptSchema = z.object({ record: z.string().max(512000), signature: z.string().regex(/^[a-f0-9]{64}$/), save_available: z.boolean().default(false) });
const sideSchema = z.object({
  gross_assets: z.number(), net_equity: z.number(), cash: z.number(), margin: z.number(),
  leverage: z.number(), largest_position_weight: z.number(), annual_volatility: z.number().nullable(),
  var_1d_95_usd: z.number().nullable(), cvar_1d_95_usd: z.number().nullable(),
  option_assets: z.number().default(0), option_liabilities: z.number().default(0),
});
export const changeComparisonSchema = z.object({
  result_id: z.string().uuid(), portfolio_id: z.string().uuid(), computed_at: z.string().datetime({ offset: true }),
  snapshot_digest: z.string(), methodology_version: z.string(),
  assumptions: z.object({ expected_portfolio_id: z.string().uuid(), ticker: z.string(), amount: z.number(), proceeds: z.enum(["cash", "repay_margin"]) }),
  price_as_of: z.string(), history_start: z.string(), observations: z.number().int(),
  sources: z.record(z.string(), z.string()), baseline: sideSchema, candidate: sideSchema,
  limitations: z.array(z.string()),
  risk_method: z.enum(["historical_equity", "mixed_instant_stress"]).default("historical_equity"),
  option_quote_basis: z.string().nullable().optional(),
  replay_receipt: receiptSchema.nullable().optional(),
  scenarios: z.array(z.object({ label: z.string(), shocks: z.record(z.string(), z.number()), iv_shift: z.number(), horizon_days: z.number(),
    baseline_pnl: z.number(), candidate_pnl: z.number(), baseline_equity: z.number(), candidate_equity: z.number() })).default([]),
  option_groups: z.array(z.object({ underlying: z.string(), expiry: z.string(), name: z.string(), leg_count: z.number(),
    mark_basis_max_loss: z.number().nullable(), mark_basis_max_gain: z.number().nullable() })).default([]),
});
export type ChangeComparison = z.infer<typeof changeComparisonSchema>;
export const comparisonVerificationSummarySchema = z.object({
  verified_at: z.string().datetime({ offset: true }), inputs_match_now: z.boolean(),
  snapshot_age_seconds: z.number().int().nonnegative(), recent_capture: z.boolean(), notice: z.string(),
});
export const comparisonVerificationSchema = comparisonVerificationSummarySchema.extend({ result: changeComparisonSchema });
export type ComparisonVerification = z.infer<typeof comparisonVerificationSummarySchema>;
export const savedComparisonSummarySchema = z.object({
  plan_id: z.string().uuid(), portfolio_id: z.string().uuid(), result_id: z.string().uuid(),
  confirmed_at: z.string().datetime({ offset: true }), notice: z.string(),
});
export const savedComparisonSchema = savedComparisonSummarySchema.extend({ result: changeComparisonSchema });
export type SavedComparison = z.infer<typeof savedComparisonSummarySchema>;
export function isChangeRequest(text: string) {
  return /^(test a change|测试变更)[.!。！]?$/i.test(text.trim());
}

File: backend/app/schemas/copilot_compare.py
"""Explicit, read-only stock/ETF reduction assumptions and paired results."""

from typing import Literal
from uuid import UUID

from pydantic import AwareDatetime, BaseModel, ConfigDict, Field, FiniteFloat, StrictBool


class CompareChange(BaseModel):
    model_config = ConfigDict(extra="forbid")

    expected_portfolio_id: UUID
    ticker: str = Field(pattern=r"^[A-Z][A-Z0-9.\-]{0,11}$")
    amount: FiniteFloat = Field(gt=0, le=100_000_000, strict=True)
    proceeds: Literal["cash", "repay_margin"]


class ComparisonSide(BaseModel):
    gross_assets: FiniteFloat
    net_equity: FiniteFloat
    cash: FiniteFloat
    margin: FiniteFloat
    leverage: FiniteFloat
    largest_position_weight: FiniteFloat
    annual_volatility: FiniteFloat | None
    var_1d_95_usd: FiniteFloat | None
    cvar_1d_95_usd: FiniteFloat | None
    option_assets: FiniteFloat = 0
    option_liabilities: FiniteFloat = 0


class ComparisonReceipt(BaseModel):
    """Opaque server-authenticated snapshot, not a save/order credential."""

    model_config = ConfigDict(extra="forbid")
    record: str = Field(max_length=512000)
    signature: str = Field(pattern=r"^[a-f0-9]{64}$")
    save_available: bool = False


class PairedStress(BaseModel):
    label: str
    shocks: dict[str, FiniteFloat]
    iv_shift: FiniteFloat
    horizon_days: int = 0
    baseline_pnl: FiniteFloat
    candidate_pnl: FiniteFloat
    baseline_equity: FiniteFloat
    candidate_equity: FiniteFloat


class UnchangedOptionGroup(BaseModel):
    underlying: str
    expiry: str
    name: str
    leg_count: int
    mark_basis_max_loss: FiniteFloat | None
    mark_basis_max_gain: FiniteFloat | None


class ChangeComparison(BaseModel):
    model_config = ConfigDict(extra="forbid")

    result_id: UUID
    portfolio_id: UUID
    computed_at: AwareDatetime
    snapshot_digest: str
    methodology_version: str
    assumptions: CompareChange
    price_as_of: str
    history_start: str
    observations: int
    sources: dict[str, str]
    baseline: ComparisonSide
    candidate: ComparisonSide
    limitations: list[str]
    risk_method: Literal["historical_equity", "mixed_instant_stress"] = "historical_equity"
    scenarios: list[PairedStress] = Field(default_factory=list)
    option_groups: list[UnchangedOptionGroup] = Field(default_factory=list)
    option_quote_basis: str | None = None
    replay_receipt: ComparisonReceipt | None = None


class ReplayComparison(BaseModel):
    model_config = ConfigDict(extra="forbid")
    expected_portfolio_id: UUID
    receipt: ComparisonReceipt


class ConfirmComparison(ReplayComparison):
    confirmed: StrictBool


class SavedComparison(BaseModel):
    model_config = ConfigDict(extra="forbid")
    plan_id: UUID
    portfolio_id: UUID
    result_id: UUID
    confirmed_at: AwareDatetime
    result: ChangeComparison
    notice: str = (
        "Saved as a draft risk plan with the original signed calculation. "
        "No holdings changed. This is a historical assumption, not an order or current recommendation."
    )


class ComparisonVerification(BaseModel):
    model_config = ConfigDict(extra="forbid")
    result: ChangeComparison
    verified_at: AwareDatetime
    inputs_match_now: bool
    snapshot_age_seconds: int = Field(ge=0)
    recent_capture: bool
    notice: str = (
        "Historical calculation reproduced without fetching market data. Not a saved plan or permission to trade."
    )
